### PR TITLE
Support collection queries

### PR DESCRIPTION
**Pull request by smcgivern**
This allows collection indices to be resolved, which means (among other things) the `required_if` option will work on collection members.

### CODE REVIEW

File: lib/attributor/attribute_resolver.rb [hunk 1]
```diff
@@ -31,11 +30,16 @@ def query!(key_path, path_prefix=ROOT_PREFIX)
       _root, *path = key_path.split(SEPARATOR)
 
       # Follow the hierarchy path to the requested node and return it
-      # Example path => ["instance", "ssh_key", "name"]
```
Comment (blanquer) [inline lib/attributor/attribute_resolver.rb]: can you leave the non-collection examples comments in there too?

File: lib/attributor/attribute_resolver.rb [hunk 2]
```diff
@@ -31,11 +30,16 @@ def query!(key_path, path_prefix=ROOT_PREFIX)
...
-      # Example @data => {"instance" => { "ssh_key" => { "name" => "foobar" } }}
+      # Example path => ["instances", "at(0)", "ssh_key", "name"]
+      # Example @data => {"instances" => [{ "ssh_key" => { "name" => "foobar" } }]}
       result = path.inject(@data) do |hash, key|
         return nil if hash.nil?
-        hash.send key
+        # at(n) means we have a collection index
+        if (match = key.match(/^at\((\d+)\)$/))
```
Comment (blanquer) [inline lib/attributor/attribute_resolver.rb]: The regexp instance should be defined at the top of the class, frozen and used here.

File: spec/attribute_resolver_spec.rb [hunk 3]
```diff
@@ -66,19 +66,50 @@
...
+
+      it 'resolves the index to the correct member of the collection' do
+        subject.query(key, prefix).should be(value)
+      end
+    end
+  end
 
   context 'checking attribute conditions' do
-    let(:key) { "instance.ssh_key.name" }
+    let(:key) { "instances.at(0).ssh_key.name" }
```
Comment (blanquer) [inline spec/attribute_resolver_spec.rb]: why are these specs changed to register a collection of instances (and therefore having to use instances.at(x) ) ?
There's really no reason to modify them right? the collection case is covered above

Comment (smcgivern) [inline spec/attribute_resolver_spec.rb]: It is, but only for querying. My use case was for the conditions, so I wanted to check that the conditions worked - and this way tests both code paths in `#query`. However, given that the specs I added above pass, this probably doesn't need to change. I'll remove it if you want.

Comment (blanquer) [inline spec/attribute_resolver_spec.rb]: yeah, I think it'd be best if we don't add that piece here. One use case at a time for tests. I'd leave this piece as it was.

**Comment by blanquer** [comment]
also, could you please squash your commits into one and sign it?
(While I haven't really documented it here yet) we're trying to follow the same contributing philosophy as Praxis: https://github.com/rightscale/praxis/blob/master/CONTRIBUTING.md#sign-your-work

**Comment by smcgivern** [comment]
Squashed, other CR change made, and signed :+1:

**Comment by blanquer** [comment]
LGTM

**Pull request merged**